### PR TITLE
Added string model parsing in clingo wrapper

### PR DESCRIPTION
**Pull request by camcunningham**
A sort of addition onto my last commit - same theme. Needed to be able to create a model based on a string instead of a temp file. If there should be a better naming convention, feel free to change it up.

Again, if this functionality exists elsewhere, let me know!

### CODE REVIEW

**Comment by gfrances** [comment]
Thanks, @camcunningham ! The functionality looks good, but we try to avoid duplicating so much code whenever possible. 

In this case, looks like the PR could be easily refactored e.g. with one single method with signature
```python
def parse_model(*, symbol_mapping, filename=None, content=None)
```
then some check to make user the user provides `filename` XOR `content`, and then refactor all common code
into a method `_parse_model(...)` that takes a Python generator providing one line of the model at a time.

Do you think you'd have the time to refactor the PR along those lines?
Cheers,

**Comment by camcunningham** [comment]
@gfrances Absolutely! Makes a lot of sense - I'll make those changes tonight

**Comment by camcunningham** [comment]
@gfrances Made some of those changes. Ended up iterating through the lines without a generator - didn't know if I'd need to use one. If you still want that I'd be happy to refactor

**Comment by gfrances** [comment]
I suggested the use of a generator because it would avoid the need for loading the entire file in memory, which is what the call to `readlines` is now doing. Usually wouldn't be such a big deal, but here we should assume that the ASP models can get pretty large on problems where the grounding is huge. Let me make some closer suggestions in the code of the PR itself.

File: src/tarski/reachability/clingo_wrapper.py [hunk 1]
```diff
@@ -40,22 +40,31 @@ def run_clingo(lp):
 
     raise ExternalCommandError(f"Unknown Gringo error. Gringo exited with code {retcode}. Full error log: {errlog}")
 
+def parse_model(*, filename=None, content=None, symbol_mapping):
+    if filename and not content:
+        with open(filename, "r") as f:
+            lines = f.readlines()
```
Comment (gfrances) [inline src/tarski/reachability/clingo_wrapper.py:46]: I would just replace this line by 
```
return _parse_model((l for l in filename), symbol_mapping)
```

File: src/tarski/reachability/clingo_wrapper.py [hunk 2]
```diff
@@ -40,22 +40,31 @@ def run_clingo(lp):
...
+    elif content and not filename:
+        lines = content.splitlines()
```
Comment (gfrances) [inline src/tarski/reachability/clingo_wrapper.py:48]: Then this one by 
```
return _parse_model((l for l in content.splitlines()), symbol_mapping)
```

**Comment by camcunningham** [comment]
Makes sense - I'll make those changes. Thanks for the info on generators!

**Comment by camcunningham** [comment]
Hm. That's strange. It passed the checks before the swap to generators - did I do it wrong or is there potentially an issue with using them there?

**Comment by haz** [comment]
Seems to be issues with the grounding on the docs with the LP solver. E.g., https://travis-ci.com/github/aig-upf/tarski/jobs/510458691

@gfrances -- any chance there's issues with gringo installing on travis?

**Comment by gfrances** [comment]
Nope, it's just that [this line](https://github.com/aig-upf/tarski/pull/110/files#diff-9d7875cf5ee765d0e41eaf0a8e137fc1ce9da88346662b8e3b3f5f02ab6ad8d6R45) is iterating over the characters of the filename, not over the contents of the filename. Should be something like:
```
    with open(filename, "r") as f:
        return _parse_model((l for l in f), symbol_mapping)
```

**Comment by camcunningham** [comment]
Ah - that was dumb of me! Took your suggestion way too literally. Should be good to go now.

File: src/tarski/reachability/clingo_wrapper.py [hunk 3]
```diff
@@ -40,22 +40,29 @@ def run_clingo(lp):
 
     raise ExternalCommandError(f"Unknown Gringo error. Gringo exited with code {retcode}. Full error log: {errlog}")
 
+def parse_model(*, filename=None, content=None, symbol_mapping):
+    if filename and not content:
+        with open(filename, "r") as f:
+            return _parse_model((l for l in f), symbol_mapping)
```
Comment (haz) [inline src/tarski/reachability/clingo_wrapper.py:46]: Would this make it an iterator, or build the full list before calling the method? Might seem weird, but would just passing in `f` work? It's treated as `lines` in `_parse_model`, and may well be just the iterator you're looking for...

Comment (gfrances) [reply src/tarski/reachability/clingo_wrapper.py:46]: The `(...)` syntax creates a generator, but you're totally right that passing in `f` should work as well;
likewise, passing in `content.splitlines()` below should work too.

**Comment by camcunningham** [comment]
@gfrances Any other changes you think need to be made?

**Comment by gfrances** [comment]
I would maybe just adopt the suggestion by Chris, passing `f` and `content.splitlines()`, but otherwise I think we're good to go?

**Comment by gfrances** [comment]
Thanks @camcunningham, @haz  !

**Pull request merged**